### DISCUSSION TITLE
sniffio._impl.AsyncLibraryNotFoundError: unknown async library, or not in async context

### DISCUSSION THREAD

**Discussion by dimaqq** [pytest-dev/pytest-asyncio#442] [General]
So I wrote some code that calls:
my code -> `httpx` -> `httpcore` -> `anyio` -> `sniffio`

When I run it "in production", using `asyncio.run(start())` it's all fine.

But then I try to start it in testing, using:

```
@pytest_asyncio.fixture
async def mything(...):
    task = asyncio.create_task(mything.start())
    yield task
    task.cancel()
```

sniffio bails with `unknown async library, or not in async context`

I'm not quite sure how to debug this... is there something special about the testing event loop?

**Comment by dimaqq** [comment]
In case someone hits this: apparently it's when the task is cancelled too soon for sniffio to make the detection, I should've known that surely it's not this library fault!

**Reply by aleksandaratanasov** [reply]
And the solution is? I mean I am writing an async client based on HTTPX that uses an AsyncIO event loop to handle sending and retrieving info. Since I am at the beginning, I cannot artificially stall the call so I get this error whenever I call my `__del__(self)`, where I call the loop to close the HTTPX async client and then stop the loop.
